NativeAOT-LLVM: start support for CALL to user functions

yowl opened the pull request:
This PR adds the beginning of support for `CALL` to user functions.  I've only done functions signatures that don't require anything passed on the shadow stack.  Have added some support for GT_LCL_VAR and GT_STORE_LCL_VAR.

Added support for ReadyToRunHelperId.GetThreadStaticBase from clrjit

If you think these PRs are a bit short or long then let me know and I'll adjust what goes in each one.

Depends https://github.com/dotnet/runtimelab/pull/945

yowl committed:
Add call helper for gc/nongc statics node

Add fallthrough branch
Add RhpAssignRef for STOREIND

yowl committed:
remove reinterpret_case

yowl committed:
remove other reinterpret_casts

yowl committed:
tidy debug and ws

yowl committed:
add support for string constants

yowl committed:
remove some debug.  make comment more confident.

yowl committed:
GT_STORE_LCL_VAR

yowl committed:
Merge branch 'cnsint-stringrefs' into stlclvar

yowl committed:
add some initial support for GT_LCL_VAR and GT_CALL for user simple signature functions (void (void/int32))

yowl committed:
consolidate get llvm sig from method

exclude methods with hidden param for now.

yowl committed:
port some pointer casting for storing simple value types

kant2002 commented on samples/HelloWorld/nuget.config:
```diff
@@ -4,6 +4,6 @@
     <clear />
     <add key="dotnet-public" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-public/nuget/v3/index.json" />
-    <add key="dotnet-experimental" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-experimental/nuget/v3/index.json" />
+    <add key="local" value="e:\github\runtimelab\artifacts\packages\Debug\Shipping" />
```
@yowl just nitpicking.

yowl replied:
Thanks

yowl committed:
revert unwanted nuget.config change

yowl committed:
cherry pick cmake fixes from commit hash 1e27105e5bd12fc8457962e497a31fcaddcd307d

yowl committed:
add helper for thread static.  Do not create duplicate LLVM functions for tentative nodes.

yowl committed:
fix test for lazy cctor presence

yowl committed:
Update llvm.cpp

Typo

jkotas commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -175,6 +213,70 @@ Value* buildAdd(llvm::IRBuilder<>& builder, GenTree* node, Value* op1, Value* op
     return mapTreeIdValue(node->gtTreeID, builder.CreateGEP(op1, op2));
 }
 
+FunctionType* getFunctionSignature(CORINFO_CONST_LOOKUP callTarget)
```
Unused?

jkotas commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -175,6 +213,70 @@ Value* buildAdd(llvm::IRBuilder<>& builder, GenTree* node, Value* op1, Value* op
+    // shadowstack arg first
+    argVec.push_back(_function->getArg(0));
+
+    // TODO Is Args() more appropriate than Operands() ?
```
I am not sure. @AndyAyersMS What is the difference between gtCall `Args()` vs `Operands()`?

AndyAyersMS replied:
`Operands()` is a superset of `Args()`, it gives you (in order)
* `this` pointer, if any;
* `Args()`
* `LateArgs()`
* call cookie (for CT_INDIRECT calls)
* call address (for CT_INDIRECT calls)
* control expression (for other sorts of indirect calls)
 
See 
https://github.com/dotnet/runtime/blob/ab820d7a68b777a2da3dd1d89d4018e9e3d3a0fe/src/coreclr/jit/compiler.hpp#L4432-L4472

jkotas commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -175,6 +213,70 @@ Value* buildAdd(llvm::IRBuilder<>& builder, GenTree* node, Value* op1, Value* op
+llvm::Value* buildUserFuncCall(GenTreeCall* gtCall, llvm::IRBuilder<>& builder)
+{
+    const char* symbolName = (*_getMangledSymbolName)(_thisPtr, gtCall->gtEntryPoint.handle);
+    // TODO: how to detect RuntimeImport ?
```
A new JIT/EE interface flag or API

yowl replied:
I did this as a another method passed in via registerLlvmCallbacks.

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -175,6 +213,70 @@ Value* buildAdd(llvm::IRBuilder<>& builder, GenTree* node, Value* op1, Value* op
+    return FunctionType::get(Type::getInt8PtrTy(_llvmContext), ArrayRef<Type*>(Type::getInt8PtrTy(_llvmContext)), false);
+}
+
+Value* genTreeAsLlvmType(GenTree* gt, Type* t)
```
Nit: any reason not to use `tree`, `type` and `value` as variable names?

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -175,6 +213,70 @@ Value* buildAdd(llvm::IRBuilder<>& builder, GenTree* node, Value* op1, Value* op
+    if (v->getType() == t)
+        return v;
+
+    if (gt->OperGet() == GT_CNS_INT && gt->gtType == var_types::TYP_INT)
```
Nit: `gt->IsIntegralConst() && gt->TypeIs(TYP_INT)`.

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -175,6 +213,70 @@ Value* buildAdd(llvm::IRBuilder<>& builder, GenTree* node, Value* op1, Value* op
+
+    if (gt->OperGet() == GT_CNS_INT && gt->gtType == var_types::TYP_INT)
+    {
+        return _builder->getInt({(unsigned int)t->getPrimitiveSizeInBits().getFixedSize(), (uint64_t)gt->AsIntCon()->gtIconVal, true});
```
Nit: `gtIconVal` -> `IconValue()`.

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -221,23 +327,80 @@ Value* buildCnsInt(llvm::IRBuilder<>& builder, GenTree* node)
     failFunctionCompilation();
 }
 
+Type*  getLLVMTypeForVarType(var_types type)
```
Nit: double space after `Type*`?

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -221,23 +327,80 @@ Value* buildCnsInt(llvm::IRBuilder<>& builder, GenTree* node)
+    builder.CreateStore(castIfNecessary(builder, toStore, getLLVMTypeForVarType(type)), castToPointerToVarType(builder, address, type));
+}
+
 void importStoreInd(llvm::IRBuilder<>& builder, GenTree* node)
```
Nit: make `node` a `GenTreeStoreInd*` and cast at the callsite? You can then use `Addr()` and `Data()` getters instead of `gtOp1` and `gtOp2`.

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -221,23 +327,80 @@ Value* buildCnsInt(llvm::IRBuilder<>& builder, GenTree* node)
 
-    // RhpAssignRef will never reverse PInvoke, so do not need to store the shadow stack here
-    builder.CreateCall(getOrCreateRhpAssignRef(), ArrayRef<Value*>{address, castIfNecessary(builder, toStore, Type::getInt8PtrTy(_llvmContext))});
+Value* localVar(llvm::IRBuilder<>& builder, GenTree* tree)
```
Nit, ditto: make `tree` a `GenTreeLclVar`?

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -221,23 +327,80 @@ Value* buildCnsInt(llvm::IRBuilder<>& builder, GenTree* node)
 }
 
-Value* visitNode(llvm::IRBuilder<> &builder, GenTree* node)
+Value* storeLocalVar(llvm::IRBuilder<>& builder, GenTree* tree)
```
Nit, ditto: make `tree` `GenTreeLclVar`?

yowl committed:
address feedback - add method isRuntimeImport

yowl committed:
Merge branch 'stlclvar' of https://github.com/yowl/runtimelab into stlclvar

yowl commented:
@SingleAccretion @jkotas Thanks for the feedback

yowl committed:
use a blittable type for the return value Value

revert gtCall->Args to gtCall->Operands() as "this" might be useful and no good reason to change it
messed up my testing for the previous commit

jkotas reviewed (approved):
LGTM
